### PR TITLE
Feature/registration page functionality

### LINKED ISSUE #1422: Customer and non customer flows
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1419

### Description

n/a

### LINKED ISSUE #1423: Form functionality
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1419

### Description

n/a

### LINKED ISSUE #1424: E2E tests of registration page
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1419

### Description

n/a

### PR DESCRIPTION
**Pull request by stevenbister**
### 🧹 Please check off the following before submitting:

- [x] Close issues
- [x] Request reviewers
- [x] Assign yourself
- [ ] Add labels (e.g. `build:osc-ecommerce`)
- [x] Add project
- [x] Set project status
- [x] Add milestone

---

## ☑ List all github issues that this PR will close:

- Closes #1422 
- Closes #1423 
- Closes #1424

## 📝 Description

Adds the registration & complete functionality to Ohana. This covers registering as a new student, completing student registration, completing tutor registration and completing admin registration.

When a user is created -- either a student by registering on the registration page, or an admin or a tutor being created elsewhere -- an email will be sent to their email address containing a link back to the registration page with a magic token which will be used to authorise them.

> [!NOTE]  
> There is currently no functionality in the api to send a magic link when an admin user is created.

When the user visits the authorised url they will have access to set their password, once they've done this they will be logged in.

The SEN section for student registration will need finishing off once the tags etc. have been merged
image
- See: #1597

## ⛳️ Current behaviour (updates)

N/a

## 🚀 New behaviour

- Adds student registration functionality
- Adds complete user registration functionality (for student/admin)
- Adds complete tutor registration functionality
- Refactors `useAuthErrors` to use generic types
- Adds zod schema and validation for registration forms
- Adds data json for registration content

## 💣 Is this a breaking change (Yes/No): No

N/a

## 🧾 Final Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible cross-device impacts (iOS, Android)
- [x] I've considered possible cross-browser impacts (Chrome, Safari, Edge, Firefox)
- [x] I've added tests to cover my changes (Unit, E2E, Integration, Contract, Functional, Acceptance etc)
- [ ] I've added or updated any documentation.

## 📝 Additional Information

You will need to add `ACADEMIC_HUB_REGISTER_PR_ROUTE=register` to the `.env` in `osc-api-auth`

When testing you need to create a user with your osc email address as Hubspot sandbox will only send to these emails. My advice would be to update your email address to something like `steve.bister+admin@openstudycollege.com` or whatever before you try registering a new user, otherwise you will get an error saying that your user already exists. Similarly if you try signing up with `steve.bister+admin@openstudycollege.com` then Hubspot won't send the email.

To do this you can run `npx prisma studio` which will give you access to the db, where you can manually change things.

In order to create a user (other than a student, which can be done on `/register`) you will need to go to `localhost:4000/graphql` and run the relevant query
> [!NOTE]  
> You will need to login and set your authorisation headers first

### Create admin
```gql
mutation CreateUser($input: createUserInput!) {
  createUser(input: $input) {
    firstName
    lastName
    email
  }
}
```

### Create tutor
```gql
mutation CreateTutor($input: createTutorInput) {
  createTutor(input: $input) {
    id
    firstName
    lastName
    email
  }
}
```

### Summary by CodeRabbit

- New Feature: Added comprehensive user registration process including token verification, form validation, and role-based form fields.
- New Feature: Enhanced user profile to display full name, email, and associated courses for tutors.
- Refactor: Improved error handling in authentication components with the introduction of `useValidationErrors` hook.
- Refactor: Updated GraphQL queries and mutations for user registration and course retrieval.
- Test: Expanded test coverage for registration forms and user-related utilities.
- Chore: Introduced new constants for user roles and updated types for form errors and registration actions.

### CODE REVIEW

File: packages/osc-academic-hub/app/utils/user.spec.ts [hunk 1]
```diff
@@ -1,30 +1,107 @@
+import { ROlES } from '~/constants';
 import type { User } from '~/types/__generated__/graphql';
+import { fullName, getUserRoles, hasRole } from './user';
 import { getUser } from './user.server';
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/utils/user.spec.ts:4]: This is incorrect, the spelling it correct

[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/utils/user.spec.ts:4]: Ahh actually my bad, I just noticed the casing issue!

File: packages/osc-academic-hub/app/routes/register.tsx [hunk 2]
```diff
@@ -32,11 +170,14 @@
...
+    if (token) {
+        const validUser = await verifyRegistrationToken(token);
 
-    if (getTokenFromUrl) {
         return json<LoaderData>({
-            token: getTokenFromUrl,
+            user: (validUser?.verifyRegistrationToken as User) ?? undefined,
+            token,
         });
     }
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/routes/register.tsx:183]: This is fine, I'm expecting the route error boundary in Remix to catch this and display the error message for me

File: packages/osc-academic-hub/app/types/general.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export interface FormErrorMessage {
+    errors?: {
+        message: string;
+    };
+}
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/types/general.ts:5]: This is fine as is, the message property will return a stringified object

**Review by mattwendzina — COMMENTED** [review]
Wow, a lot of work here..! Nice work mate. I've not gone through in major detail, but it all seems to looks good other than the few error issues you've been getting which we've discussed.
One other thought/thing I've noticed. When you register, it still just sits on the same page and you can click sign up again (See image). It doesn't let you because it recognises a user has been created, but I think it would be nice to just hide the form once they completed the sign up. 
Screenshot 2023-09-12 at 14 50 22

**Comment by stevenbister** [comment]
> Wow, a lot of work here..! Nice work mate. I've not gone through in major detail, but it all seems to looks good other than the few error issues you've been getting which we've discussed.

Aha thanks man, yeah these last few tasks have been pretty meaty 😅 (and this doesn't include the styling/setup parts lol) I think I've solved the issues you were having around session size here: https://github.com/Open-Study-College/osc/pull/1615

> One other thought/thing I've noticed. When you register, it still just sits on the same page and you can click sign up again (See image). It doesn't let you because it recognises a user has been created, but I think it would be nice to just hide the form once they completed the sign up.

Ooo yeah that's a good shout, I'll have a play around and see if I can come up with something nice

**Comment by stevenbister** [comment]
> Wow, a lot of work here..! Nice work mate. I've not gone through in major detail, but it all seems to looks good other than the few error issues you've been getting which we've discussed. One other thought/thing I've noticed. When you register, it still just sits on the same page and you can click sign up again (See image). It doesn't let you because it recognises a user has been created, but I think it would be nice to just hide the form once they completed the sign up. Screenshot 2023-09-12 at 14 50 22

Okay I've set it to remove the form when there is a success message and that should just display the alert. At the moment it looks a bit weird being that small but I've added a large variant to #1605 which looks like this
image
(Ignore the text lol)

So we'll be able to replace it once that goes in and it should look a bit more balanced.

Would be nice to add some micro interactions or little animations a bit further down the line too

**Comment by mattwendzina** [comment]
> > Wow, a lot of work here..! Nice work mate. I've not gone through in major detail, but it all seems to looks good other than the few error issues you've been getting which we've discussed. One other thought/thing I've noticed. When you register, it still just sits on the same page and you can click sign up again (See image). It doesn't let you because it recognises a user has been created, but I think it would be nice to just hide the form once they completed the sign up. Screenshot 2023-09-12 at 14 50 22
> 
> Okay I've set it to remove the form when there is a success message and that should just display the alert. At the moment it looks a bit weird being that small but I've added a large variant to #1605 which looks like this image (Ignore the text lol)
> 
> So we'll be able to replace it once that goes in and it should look a bit more balanced.
> 
> Would be nice to add some micro interactions or little animations a bit further down the line too

Like it mate. Only thing, in terms of the message, wonder if it's worth adding something like... 'If you don't receive an email please contact us at openstudycollege.com or phone us at ...'

**Review by mattwendzina — COMMENTED** [review]
I've left one follow up comment above in regards to the registration success message. Other than that though, looks good to go 🚢

**Review by Tom-Pearce — COMMENTED** [review]
I've not looked at the code, purely from a functionality perspective.

Going from the login page to the `/register` page (clicking "Dont have an account?") is a very jarring experience. Everything along a vertical mirror (image moves to the left, form to the right), the background changes to a gradient thats barely perceivable so just looks plain and an after thought. The image is no longer inside the card, but just hovers outside to the left.

Discussed this with @stevenbister, maybe @hi-im-si  has some thoughts; but my suggestion is to remain on the main login page, but just scroll the `/login` form out to the left, and scroll the `/register` form in from the right. You could even go as far as scrolling _everything_ (image included) left, and having the registration form on the right again - at least in this scenario it feels smoother and is not jarring at all.

My main gripe was when I first saw the registration page it took me a good few seconds to process what I was looking at since it all moved so much and so suddenly (no transition). Other sign up experiences I trialled with Steven were much smoother and the page remained much more consistent with the login page.

Only other point is I think once you complete /registration you should be redirected to the /dashboard and not the /profile

**Comment by hi-im-si** [comment]
> I've not looked at the code, purely from a functionality perspective.
> 
> Going from the login page to the `/register` page (clicking "Dont have an account?") is a very jarring experience. Everything along a vertical mirror (image moves to the left, form to the right), the background changes to a gradient thats barely perceivable so just looks plain and an after thought. The image is no longer inside the card, but just hovers outside to the left.


I get what you mean @Tom-Pearce but i think this may be an adjustment that we may need to sacrifice at this crucial stage of the project. I think we all had a good chance to comment on the design of these pages in figma and some of these suggestions effectively are a complete change to the registration page design.....that said, i do agree that the difference between the pages is a little unconventional - i guess if these pages are to look different then there needs to be some rationale behind it.


> Discussed this with @stevenbister, maybe @hi-im-si has some thoughts; but my suggestion is to remain on the main login page, but just scroll the `/login` form out to the left, and scroll the `/register` form in from the right. You could even go as far as scrolling _everything_ (image included) left, and having the registration form on the right again - at least in this scenario it feels smoother and is not jarring at all.

As before, this is really going to be down to how significant of an update this is. time is not on our side currently. @stevenbister may be able to best advise here

> 
> My main gripe was when I first saw the registration page it took me a good few seconds to process what I was looking at since it all moved so much and so suddenly (no transition). Other sign up experiences I trialled with Steven were much smoother and the page remained much more consistent with the login page.

As mentioned, i agree, these pages tend to stay fairly similar in appearance but again this si something we a) should have picked up sooner and b) not sure if we are afforded the time to update.

> 
> Only other point is I think once you complete /registration you should be redirected to the /dashboard and not the /profile

I agree with this

**Comment by hi-im-si** [comment]
Points from me:
- Upon successfully registering, it might be worth having a link to go to login or to go back to our website
- Upon registering and then refreshing the page, i attempted to register again but received the following error:
image
What came back in the terminal was the following:

```bash
UnauthorizedError: jwt expired
    at new UnauthorizedError (/Users/simonshahriveri/Desktop/SITES/__OSC/osc-mono/node_modules/express-jwt/dist/errors/UnauthorizedError.js:22:28)
    at /Users/simonshahriveri/Desktop/SITES/__OSC/osc-mono/node_modules/express-jwt/dist/index.js:133:35
    at step (/Users/simonshahriveri/Desktop/SITES/__OSC/osc-mono/node_modules/express-jwt/dist/index.js:33:23)
    at Object.next (/Users/simonshahriveri/Desktop/SITES/__OSC/osc-mono/node_modules/express-jwt/dist/index.js:14:53)
    at fulfilled (/Users/simonshahriveri/Desktop/SITES/__OSC/osc-mono/node_modules/express-jwt/dist/index.js:5:58)
{
  "statusId": "Wfb4WWZBQeeM7wHvV9l4HgAAAYqoL6-OAZSpCQ==",
  "requestedAt": "2023-09-18T12:06:52.132Z",
  "status": "PENDING"
}
```

**Comment by hi-im-si** [comment]
Forgive me if this is a known issue but upon registering and then receiving the link to my email.
The link directs me to the following url:
http://localhost:3000/register?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyIjp7ImlkIjo0OX0sImlhdCI6MTY5NTAzOTU4NiwiZXhwIjoxNjk1MDUwMzg2LCJhdWQiOiJodHRwOi8vbG9jYWxob3N0OjQwMDAvZ3JhcGhxbCJ9.Qwf4V1iPDITKoMYXNrl4YkaXH8K-obIPJA7mwKg72wM&utm_medium=email&_hsmi=71915580&_hsenc=p2ANqtz-86xG1vpLJyYUqEKbLHM2DPFZ1J8-28bUCXBDB4x09dA-GLLYHxEGuh-FQGctumSCtvmDNcSNC1QZk8dBiNYN9nZs7Qd_G5IB38Xesl9YHDJ0zsi4E&utm_content=71915580&utm_source=hs_email

But this returned the following error page:
image

Once i removed the following params from the URL: `&utm_source=hs_email` - this then correctly directed me to the "complete registration" page as expected.
@Tom-Pearce / @hall2521 / @stevenbister - something to be aware of!

**Comment by Tom-Pearce** [comment]
> > Only other point is I think once you complete /registration you should be redirected to the /dashboard and not the /profile
> 
> I agree with this but then we may also want to encourage the user to update their profile?

Arguably this should be done as part of `/register` when they're setting their password. I don't think there should be any data the user needs to update that shouldn't be part of the registration process.

Completely agree about the time constraints for the layout, although I do think that as this is possibly the _very_ first experience any user may have the page should create a better first impression than it currently does, so maybe this is something worth putting time into to tweak the layout; even if it just replicates the login card layout. And it's very hard to see how one layout flows into the next in figma, as they're just sat their together on the same screen - there was no flow/presentation setup for us to view it properly one page after another.

**Comment by stevenbister** [comment]
> Forgive me if this is a known issue but upon registering and then receiving the link to my email. The link directs me to the following url: http://localhost:3000/register?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyIjp7ImlkIjo0OX0sImlhdCI6MTY5NTAzOTU4NiwiZXhwIjoxNjk1MDUwMzg2LCJhdWQiOiJodHRwOi8vbG9jYWxob3N0OjQwMDAvZ3JhcGhxbCJ9.Qwf4V1iPDITKoMYXNrl4YkaXH8K-obIPJA7mwKg72wM&utm_medium=email&_hsmi=71915580&_hsenc=p2ANqtz-86xG1vpLJyYUqEKbLHM2DPFZ1J8-28bUCXBDB4x09dA-GLLYHxEGuh-FQGctumSCtvmDNcSNC1QZk8dBiNYN9nZs7Qd_G5IB38Xesl9YHDJ0zsi4E&utm_content=71915580&utm_source=hs_email
> 
> But this returned the following error page: image
> 
> Once i removed the following params from the URL: `&utm_source=hs_email` - this then correctly directed me to the "complete registration" page as expected. @Tom-Pearce / @hall2521 / @stevenbister - something to be aware of!

This is a thing with Brave. It strips the `http` part out of the url, presumably because it's not `https`. Works fine everywhere else as far as I've tested

File: packages/osc-api-auth/src/types/functions.ts [hunk 4]
```diff
@@ -47,13 +55,8 @@ export type PasswordResetFn = (
 export type UpdatePasswordFn = (input: passwordUpdateInput) => Promise<User | Error>;
 export type UpdatePasswordFromOrderFn = (input: passwordResetInput) => Promise<User | Error>;
```
[the remark this answers is not in the document]
Comment (Tom-Pearce) [reply packages/osc-api-auth/src/types/functions.ts:56]: @coderabbitai, these errors are returned within the context of Apollo GraphQL and GraphQL Shield which handles the errors appropriately and outputs them in line with GQL standards

**Comment by stevenbister** [comment]
> As before, this is really going to be down to how significant of an update this is. time is not on our side currently. @stevenbister may be able to best advise here
> 
> > My main gripe was when I first saw the registration page it took me a good few seconds to process what I was looking at since it all moved so much and so suddenly (no transition). Other sign up experiences I trialled with Steven were much smoother and the page remained much more consistent with the login page.
> 
> As mentioned, i agree, these pages tend to stay fairly similar in appearance but again this si something we a) should have picked up sooner and b) not sure if we are afforded the time to update.

I do get what's being said, and admittedly it's not a massive piece of work (well the animation bit is a bit more :P)
But I think given the time scales and the amount of other stuff I still need to get through I'd be inclined to leave it as is for now and focus on getting these other little bugs squashed and the rest of the work moved forward. Not sure if it's been said above but perhaps something I can pick up after we go live. I wanna add in some animations and little micro-interactions at some point anyway so might be a good time to do this as well.

> > Only other point is I think once you complete /registration you should be redirected to the /dashboard and not the /profile
> 
> I agree with this but then we may also want to encourage the user to update their profile?

Yeah I went back and forth on this for a bit, given we don't really have a dashboard or anything else yet I figured directing to the profile was the best bet as originally that was going to be the next piece to go in. That's changed now though so I'll point them at the dashboard and just list out the courses and profile info on there for the time being.

**Comment by hi-im-si** [comment]
> > Forgive me if this is a known issue but upon registering and then receiving the link to my email. The link directs me to the following url: http://localhost:3000/register?token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VyIjp7ImlkIjo0OX0sImlhdCI6MTY5NTAzOTU4NiwiZXhwIjoxNjk1MDUwMzg2LCJhdWQiOiJodHRwOi8vbG9jYWxob3N0OjQwMDAvZ3JhcGhxbCJ9.Qwf4V1iPDITKoMYXNrl4YkaXH8K-obIPJA7mwKg72wM&utm_medium=email&_hsmi=71915580&_hsenc=p2ANqtz-86xG1vpLJyYUqEKbLHM2DPFZ1J8-28bUCXBDB4x09dA-GLLYHxEGuh-FQGctumSCtvmDNcSNC1QZk8dBiNYN9nZs7Qd_G5IB38Xesl9YHDJ0zsi4E&utm_content=71915580&utm_source=hs_email
> > But this returned the following error page: image
> > Once i removed the following params from the URL: `&utm_source=hs_email` - this then correctly directed me to the "complete registration" page as expected. @Tom-Pearce / @hall2521 / @stevenbister - something to be aware of!
> 
> This is a thing with Brave. It strips the `http` part out of the url, presumably because it's not `https`. Works fine everywhere else as far as I've tested

Bloody brave 🤦‍♂️

**Comment by hi-im-si** [comment]
> > > Only other point is I think once you complete /registration you should be redirected to the /dashboard and not the /profile
> > 
> > 
> > I agree with this but then we may also want to encourage the user to update their profile?
> 
> Arguably this should be done as part of `/register` when they're setting their password. I don't think there should be any data the user needs to update that shouldn't be part of the registration process.
> 
> Completely agree about the time constraints for the layout, although I do think that as this is possibly the _very_ first experience any user may have the page should create a better first impression than it currently does, so maybe this is something worth putting time into to tweak the layout; even if it just replicates the login card layout. And it's very hard to see how one layout flows into the next in figma, as they're just sat their together on the same screen - there was no flow/presentation setup for us to view it properly one page after another.

More than happy for us to point straight to dashboard - its pretty standard i guess. I think we wanted to ensure only the vital information was processed at registration - its important from a usability pov that we onboard customers/users as quickly as possible. The extra information such as Bio, Avatar etc are things we would like them to fill out but not essential. 
As for the login/registration - I think the layouts were pretty obvious from the beginning on their differences - i guess ultimately none of us raised them. As discussed with @stevenbister we will leave things as they are for the moment - time is not on our side. It will be great to tackle these once everything else has been completed and if we decide to limit who sees AH in the early days then it might not be too much of a problem. @stevenbister def worth logging as an issue for post-live.

**Comment by stevenbister** [comment]
@hi-im-si @Tom-Pearce Issues above should be sorted. Let me know if I've missed any got a bit hard to follow with the rabbit popping up in between 😂 

I've also added the course confirmation bit to the tutor registration you mentioned, so now if you're a tutor you will see a list of checkboxes that show which courses they've been assigned to. They can uncheck them to not agree to those and it shouldn't add them to their user once they confirm.

File: packages/osc-academic-hub/app/types/__generated__/gql.ts [hunk 5]
```diff
@@ -53,7 +59,31 @@ export function gql(source: "\n    mutation CompleteResetPassword($input: passwo
...
+/**
+ * The gql function is used to parse GraphQL queries into a document that can be used by GraphQL clients.
+ */
+export function gql(source: "\n    query TutorCourses($tutorCoursesId: Int) {\n        tutorCourses(id: $tutorCoursesId) {\n            course {\n                title\n            }\n        }\n    }\n"): (typeof documents)["\n    query TutorCourses($tutorCoursesId: Int) {\n        tutorCourses(id: $tutorCoursesId) {\n            course {\n                title\n            }\n        }\n    }\n"];
 
 export function gql(source: string) {
```
[the remark this answers is not in the document]
Comment (stevenbister) [reply packages/osc-academic-hub/app/types/__generated__/gql.ts:88]: @coderabbitai don't review these files, they're auto generated so we don't have control over the output

**Review by hi-im-si — APPROVED** [review]

**Pull request merged**